uint16 wls1
uint16 wls2
uint16 wls3